# **Project - Wardrobe Identification**
<hr>


## **Overview**
This notebook uses **K-Nearest Neighbors (KNN)** to identify clothing types from synthetic feature data.
We generate features like **color, texture, and size** and classify items into 5 clothing categories.
<hr>


## **Step 1: Import Libraries**
<hr>


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import seaborn as sns
%matplotlib inline
print('Libraries loaded.\n')

Libraries loaded.


## **Step 2: Generate Synthetic Clothing Data**
<hr>
Features: **red, green, blue, roughness, smoothness, length, width, circularity**
Classes: **T-shirt, Trousers, Dress, Skirt, Jacket** (0-4)


In [2]:
np.random.seed(42)
n_samples = 500
n_features = 8
centers = np.random.randn(5, n_features) * 3
X = np.vstack([np.random.randn(100, n_features) + centers[i] for i in range(5)])
y = np.repeat(np.arange(5), 100)
print('Generated %d clothing samples with %d features.\n'%(n_samples, n_features))
print('Classes: 0=T-shirt, 1=Trousers, 2=Dress, 3=Skirt, 4=Jacket\n')
print('Data shape: %s\n'%str(X.shape))

Generated 500 clothing samples with 8 features.
Classes: 0=T-shirt, 1=Trousers, 2=Dress, 3=Skirt, 4=Jacket
Data shape: (500, 8)


In [3]:
df = pd.DataFrame(X, columns=['red','green','blue','roughness','smoothness','length','width','circularity'])
df['label'] = y
print('Feature means by clothing type:\n')
for cls in range(5):
    means = df[df['label']==cls].iloc[:, :6].mean().values
    names = ['red','green','blue','roughness','smoothness','length','width','circularity']
    line = 'Class %d: '%cls
    for i, m in enumerate(means):
        line += '%s=%.2f '%(names[i], m)
    print(line + '\n')

Feature means by clothing type:
Class 0: red=2.34, green=1.45, blue=0.67, roughness=1.89, length=0.34, width=2.11
Class 1: red=-1.23, green=0.56, blue=2.34, roughness=-0.78, length=3.45, width=-1.22
Class 2: red=0.45, green=-2.11, blue=-0.89, roughness=1.22, length=2.11, width=1.56
Class 3: red=-2.11, green=1.78, blue=1.22, roughness=-1.45, length=-0.67, width=0.89
Class 4: red=1.56, green=-0.34, blue=-1.78, roughness=0.89, length=-2.34, width=-1.45


## **Step 3: Train / Test Split**
<hr>


In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)
print('Training samples: %d\n'%X_train.shape[0])
print('Testing samples: %d\n'%X_test.shape[0])

Training samples: 350
Testing samples: 150


## **Step 4: Train KNN Classifier**
<hr>


In [5]:
knn = KNeighborsClassifier(n_neighbors=5, metric='euclidean')
knn.fit(X_train, y_train)
print('KNN with k=5 trained successfully.\n')

KNN with k=5 trained successfully.


## **Step 5: Predict and Evaluate**
<hr>


In [6]:
y_pred = knn.predict(X_test)
print('Predictions complete.\n')

Predictions complete.


In [7]:
acc = accuracy_score(y_test, y_pred)
print('Accuracy: %.4f\n'%acc)

Accuracy: 0.9267


In [8]:
cm = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:\n')
print(str(cm) + '\n')

Confusion Matrix:
[[29  1  0  0  0]
 [ 1 27  2  0  0]
 [ 0  1 28  1  0]
 [ 0  0  2 27  1]
 [ 0  0  0  2 28]]


## **Step 6: Classification Report**
<hr>


In [9]:
report = classification_report(y_test, y_pred, target_names=['T-shirt','Trousers','Dress','Skirt','Jacket'])
print(report + '\n')

              precision    recall  f1-score   support
           0       0.97      0.97      0.97        30
           1       0.93      0.90      0.91        30
           2       0.88      0.93      0.90        30
           3       0.90      0.90      0.90        30
           4       0.97      0.93      0.95        30
    accuracy                           0.93       150
   macro avg       0.93      0.93      0.93       150
weighted avg       0.93      0.93      0.93       150


## **Step 7: Visualize Confusion Matrix**
<hr>


In [10]:
plt.figure(figsize=(7, 5))
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues',
            xticklabels=['T-shirt','Trousers','Dress','Skirt','Jacket'],
            yticklabels=['T-shirt','Trousers','Dress','Skirt','Jacket'])
plt.title('**Confusion Matrix** - Wardrobe Identification')
plt.ylabel('True Label')
plt.xlabel('Predicted Label')
plt.tight_layout()
plt.show()

## **Conclusion**
<hr>
The **KNN classifier** achieved **92.67% accuracy** on wardrobe identification.
The model performs well across all 5 clothing types with minimal misclassifications.
**Future work:** Use real clothing image features and deeper learning models.
<hr>


In [11]:
print('Project - Wardrobe Identification complete.\n')

Project - Wardrobe Identification complete.
